ISPN-13834 Create single REST endpoint for all container events

jabolina opened the pull request:
https://issues.redhat.com/browse/ISPN-13834

* Created an endpoint to listen for lifecycle events at
`/rest/v2/container?action=listen`. The old endpoint still exists. We
listen for logs produced with the `ServerEventLogger`.

* The current endpoint only publishes events logged with LIFECYCLE
category.

* Currently we only publish the event in the JSON format. We need more
work to handle properly the `Accept` HTTP header.

* Currently we publish all events received by the logger. There is room
for future improvement to listen only events for the node which is
serving the request.

jabolina commented:
I've created https://issues.redhat.com/browse/ISPN-13873 to implement the event serialization correctly. I am opening this for review with this limitation.

ryanemerson commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -533,13 +547,31 @@ private CompletionStage<RestResponse> shutdown(RestRequest restRequest) {
+      final CompletionStage<Void> cs = cacheManager.addListenerAsync(listener);
+      if (checkEvents) {
+         InternalCacheRegistry icr = registry.getComponent(InternalCacheRegistry.class);
+         Cache<?, ?> cache = icr.retrieveInternalCache(SERVER_LOGGER_CACHE_NAME);
```
I don't think we need this new `InternalCacheRegistry` method. Instead we can do:

```java
         Cache<?, ?> cache = cacheManager.getCache(SERVER_LOGGER_CACHE_NAME);
         cs.thenCompose(v -> cache.addListenerAsync(listener));
```

With ^ I think a `null` check is redundant as the cache is always created on cache manager startup, so by the time a pod is able to serve requests the cache should always be available.

jabolina replied:
I did that first one with the registry because I thought that the events cache was internal, but it works correctly using the cache manager directly. Should we create if it not exists? Because the `ServerEventLogger` should handle this, and if it was not created there probably means that we are using another kind of logger.

ryanemerson commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -533,13 +547,31 @@ private CompletionStage<RestResponse> shutdown(RestRequest restRequest) {
-      return cacheManager.addListenerAsync(listener).thenApply(v -> responseBuilder.build());
+
+      final CompletionStage<Void> cs = cacheManager.addListenerAsync(listener);
+      if (checkEvents) {
```
Nitpick, but maybe `includeEvents` is more descriptive?

ryanemerson commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -578,6 +610,19 @@ public EventStream getEventStream() {
       }
 
+      @CacheEntryCreated
+      public CompletionStage<Void> onEntryCreated(CacheEntryCreatedEvent<?, EventLog> entry) {
```
As the cache always has `UUID` as a key, I think we can replace the wildcard with `UUID` and avoid the unchecked cast.

ryanemerson commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -578,6 +610,19 @@ public EventStream getEventStream() {
 
+      @CacheEntryCreated
+      public CompletionStage<Void> onEntryCreated(CacheEntryCreatedEvent<?, EventLog> entry) {
+         if (!(entry instanceof EventImpl) || entry.isPre()) return CompletableFutures.completedNull();
```
Why does the event have to be an instance of `EventImpl`? We never use any implementation specific methods, so any `EventLog` implementation should be fine. My understanding is that we would want to include `ClusterEvent` for events that did not originate on the node that was being listened to.

jabolina replied:
Yeah, verifying here it is not really necessary. I started the initial implementation using it and just kept :/

ryanemerson commented on core/src/main/java/org/infinispan/registry/InternalCacheRegistry.java:
```diff
@@ -102,4 +103,14 @@ enum Flag {
+    * @param <V>: Cache value type.
+    * @return The internal cache if exists, null otherwise.
+    */
+   <K, V> Cache<K, V> retrieveInternalCache(String name);
```
See later comment, but I don't think this new method is necessary.

ryanemerson commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -599,4 +644,30 @@ public CompletionStage<Void> onConfigurationEvent(ConfigurationChangedEvent even
    }
+
+   @Immutable
+   private class LifecycleEvent implements JsonSerialization, Serializable {
```
Class can be `static`.

ryanemerson replied:
Is there any reason why this implements `Serializable`? In general we avoid this as we rely on ProtoStream/Externalizers for internal marshalling requirements.

jabolina replied:
Guess it was just habit in adding the `Serializable`. Updating.

ryanemerson commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -578,6 +610,19 @@ public EventStream getEventStream() {
+         EventLog event = eventImpl.getValue();
+         if (event.getCategory() != EventLogCategory.LIFECYCLE) return CompletableFutures.completedNull();
+
+         final LifecycleEvent serializable = new LifecycleEvent(event);
```
`lifecycleEvent`?

ryanemerson commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -578,6 +610,19 @@ public EventStream getEventStream() {
+         if (event.getCategory() != EventLogCategory.LIFECYCLE) return CompletableFutures.completedNull();
+
+         final LifecycleEvent serializable = new LifecycleEvent(event);
+         final ServerSentEvent sse = new ServerSentEvent("lifecycle-event", serializable.toString());
```
I think we should be sending this as `serializable.toJson()` in order to reduce the payload size.

ryanemerson commented on server/rest/src/test/java/org/infinispan/rest/resources/WeakSSEListener.java:
```diff
@@ -0,0 +1,82 @@
+
+   public void expectEvent(String type, String subString) throws InterruptedException {
+      CompletableFuture<Void> waitEvent = CompletableFuture.supplyAsync(() -> {
+         AtomicBoolean contains = new AtomicBoolean(false);
```
As we don't need concurrency control, we can just use `ByRef.Boolean(false)`

ryanemerson commented on server/rest/src/test/java/org/infinispan/rest/resources/WeakSSEListener.java:
```diff
@@ -0,0 +1,82 @@
+            for (int i = 0; i < v.size() && !contains.get(); i++) {
+               if (v.get(i).contains(subString)) {
+                  contains.set(true);
+                  index = i;
```
Should we break here? Currently if the value contains multiple strings with the same `subString` the latest will be removed providing LIFO semantics, but I would expect FIFO.

ryanemerson commented on server/rest/src/test/java/org/infinispan/rest/resources/WeakSSEListener.java:
```diff
@@ -0,0 +1,82 @@
+               } else {
+                  backup.compute(event.getKey(), compute(event));
+               }
+            } catch (InterruptedException e) {
```
This can be simplified to:

```java
               if (type.equals(event.getKey()) && event.getValue().contains(subString)) {
                  contains.set(true);
                  break;
               } else {
                  backup.compute(event.getKey(), compute(event));
               }
```

We can then just define the compute `BiFunction` inline as it's only required once.

ryanemerson reviewed (changes requested):
Great start @jabolina!

It doesn't have to be resolved in this PR, but this has made me think that all configuration events should be recorded in the `EventLog` as `LIFECYCLE` events as that is what they are. Once that's in place, it becomes possible to remove the `ConfigurationEvent#onConfigurationEvent` logic from `ContainerResource` completely. The only downside I can think is that this means a write to the `___event_log_cache` for every new cache/configuration change when in server mode.

@tristantarrant WDYT?

jabolina commented:
> Great start @jabolina!
> 
> It doesn't have to be resolved in this PR, but this has made me think that all configuration events should be recorded in the `EventLog` as `LIFECYCLE` events as that is what they are. Once that's in place, it becomes possible to remove the `ConfigurationEvent#onConfigurationEvent` logic from `ContainerResource` completely. The only downside I can think is that this means a write to the `___event_log_cache` for every new cache/configuration change when in server mode.
> 
> @tristantarrant WDYT?

One limitation by using only the events logger is that the entries are stored in the cache, so that `includeCurrentState` boolean read all persisted data in the cache. If the data persists through restarts, we would read an ever-increasing list of events. 

But I also think we should migrate some of the events to be `LIFECYCLE`. At this moment we are publishing only the rebalace events as lifecycle.

ryanemerson commented:
> One limitation by using only the events logger is that the entries are stored in the cache, so that `includeCurrentState` boolean read all persisted data in the cache. If the data persists through restarts, we would read an ever-increasing list of events.

That's a very good point! The cache is Queryable, so I guess a workaround would be if `includeCurrentState` only return events that are related to configuration, using a combination of `EventLog#getMessage` and `EventLog#getContext` to ensure a single event is sent per resource config. But that might be worth way more hassle than it's worth, with a simpler solution being to just extend `ContainerResource.ConfigurationListener` to listen to additional cache events for other resource types e.g. Protobuf schema changes, Counters etc.

tristantarrant commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -533,13 +545,30 @@ private CompletionStage<RestResponse> shutdown(RestRequest restRequest) {
+
+      final CompletionStage<Void> cs = cacheManager.addListenerAsync(listener);
+      if (includeEvents) {
+         Cache<?, ?> cache = cacheManager.getCache(SERVER_LOGGER_CACHE_NAME, false);
```
Instead of retrieving the logger cache like this (which leaks an implementation detail of the server logger), we should instead add
a `EventLogger getEventLogger();` method to `ServerManagement`.
`EventLogger` should have a way to publish events to some kind of `EventLogListener`.

jabolina replied:
I've created https://issues.redhat.com/browse/ISPN-13880 to follow-up

jabolina commented:
> @jabolina LGTM. Are you planning to address ISPN-13873 in this PR or a subsequent one?

I'll open a subsequent one, to avoid a large diff.

ryanemerson commented:
On hold until https://github.com/infinispan/infinispan/pull/10118 is ready and PR rebased to use the changes.

wburns commented:
> On hold until #10118 is ready and PR rebased to use the changes.

Looks like this PR can go ahead now? Is it okay for review? Can we removed the Oh hold tag?

jabolina commented:
Should be good to go now.

jabolina commented:
Last CI run did not complete, pushed with Will's suggestion to trigger CI again.

pruivo commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -533,12 +541,25 @@ private CompletionStage<RestResponse> shutdown(RestRequest restRequest) {
       responseBuilder.contentType(TEXT_EVENT_STREAM).entity(listener.getEventStream());
+
+      if (includeLifecycle) {
+         EventLogManager.getEventLogger(cacheManager).addListener(listener);
```
shouldn't use `addListenerAsync`?

pruivo commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -125,6 +129,10 @@ public Invocations getInvocations() {
 
+            // Container lifecycle listener
+            .invocation().method(GET).path("/v2/container").withAction("listen")
+               .permission(AuthorizationPermission.ADMIN).auditContext(AuditContext.SERVER).handleWith(this::listenLifecycle)
```
Are the config events supposed to be sent/received in this endpoint?

pruivo replied:
so why create a new endpoint? why not just create a new parameter in `config` stream to include the lifecycle events?

ryanemerson replied:
The reason for a new endpoint is to reflect that this isn't specific to config, but container events in general. It's not intuitive for a stream of config+lifecycle events to be consumed from a `config` sub-resource i.e .`/v2/container/config`

pruivo replied:
well, it is confusing to me. If I want only lifecycle events, I have to use `/v2/container?action=listen` and filter out the events I don't want...  
If we add view change (and xsite view change) events, would they go into this new endpoint? or create a new one?

ryanemerson replied:
I think all events should be exposed via `/v2/container`. There's no need to create a new endpoint for the events you describe. I agree that some kind of server-side filtering would be nice to allow only requested events to be sent over the wire. We could add support for filtering via parameters, with only requested event types being returned if one or more filters are applied, the default being to send everything if no filters are applied.

pruivo replied:
+100 to filtering

ryanemerson replied:
WDYT @jabolina?

This PR has been open for a while now, so I think it's acceptable to merge in it's current state (once the SecurityAction changes have been applied) and create a Jira to add event filtering in a subsequent PR.

pruivo commented:
@jabolina the documentation needs to be updated with the new endpoint/path.

pruivo commented:
there is a bug in `ServerEventImpl` class. In the log file I get
```
2022-07-14 15:57:08,178 INFO  (jgroups-6,pedro-laptop-4-14764) [org.infinispan.LIFECYCLE] [Context=org.infinispan.topology.LocalTopologyManagerImpl][Context=pedro-laptop-4-14764]ISPN100002: Starting rebalance with members [pedro-laptop-4-14764, pedro-laptop-4-1491], phase READ_OLD_WRITE_ALL, topology id 2

```

But in REST I got the following event
```
event: lifecycle-event
data: {"message":"ISPN100002: Starting rebalance with members [pedro-laptop-4-14764, pedro-laptop-4-1491], phase READ_OLD_WRITE_ALL, topology id 2","category":"LIFECYCLE","level":"INFO","time":1657810628178,"owner":"","context":"org.infinispan.topology.LocalTopologyManagerImpl","scope":"pedro-laptop-4-14764"}
```

Note that the cache name is not present in the event.

In `ServerEventImpl`, the `who` and `context` are swapped in the constructor on line 28.

jabolina commented:
> there is a bug in `ServerEventImpl` class. In the log file I get
> 
> ```
> 2022-07-14 15:57:08,178 INFO  (jgroups-6,pedro-laptop-4-14764) [org.infinispan.LIFECYCLE] [Context=org.infinispan.topology.LocalTopologyManagerImpl][Context=pedro-laptop-4-14764]ISPN100002: Starting rebalance with members [pedro-laptop-4-14764, pedro-laptop-4-1491], phase READ_OLD_WRITE_ALL, topology id 2
> ```
> 
> But in REST I got the following event
> 
> ```
> event: lifecycle-event
> data: {"message":"ISPN100002: Starting rebalance with members [pedro-laptop-4-14764, pedro-laptop-4-1491], phase READ_OLD_WRITE_ALL, topology id 2","category":"LIFECYCLE","level":"INFO","time":1657810628178,"owner":"","context":"org.infinispan.topology.LocalTopologyManagerImpl","scope":"pedro-laptop-4-14764"}
> ```
> 
> Note that the cache name is not present in the event.
> 
> In `ServerEventImpl`, the `who` and `context` are swapped in the constructor on line 28.

Good catch! I've updated the constructor so both receive the arguments in the same order and fixed the decorated event logger, everything was added as `Context` in the log. Now we have:

```text
15:52:25,194 INFO  (non-blocking-thread--p2-t12) [org.infinispan.LIFECYCLE] [Context=org.infinispan.topology.LocalTopologyManagerImpl][Scope=jbolina-15076]ISPN100002: Starting rebalance with members [jbolina-47009, jbolina-15076], phase READ_OLD_WRITE_ALL, topology id 2
```

pruivo commented:
@jabolina I actually have the wrong line in ServerEventImpl. The context should contain the cache's name and not the class :)

ryanemerson commented on server/rest/src/main/java/org/infinispan/rest/resources/ContainerResource.java:
```diff
@@ -533,13 +541,27 @@ private CompletionStage<RestResponse> shutdown(RestRequest restRequest) {
       responseBuilder.contentType(TEXT_EVENT_STREAM).entity(listener.getEventStream());
-      return cacheManager.addListenerAsync(listener).thenApply(v -> responseBuilder.build());
+
+      CompletionStage<Void> cs = cacheManager.addListenerAsync(listener);
```
I think we need to do this (plus subsequent calls) via a `SecurityAction` for when Authorization is enabled, as `addListenerAsync` checks for the LISTEN permission.

The original container SSE endpoint was missing similar checks, see https://github.com/infinispan/infinispan/pull/10232 for details.

jabolina replied:
Updated, thanks! I've created a new one for adding the listener into the logger and kind of cloned the existent one to use the async method.

jabolina commented:
> @jabolina I actually have the wrong line in ServerEventImpl. The context should contain the cache's name and not the class :)

I've updated so we use the cache name. Right now, we publish lifecycle events only in the `LocalTopologyManagerImpl` and its a global component, so we have to update the cache name on each operation. I think that we may need to update the `DecoratedEventLogger` to be immutable, but let me know what you think.

```text
13:57:03,994 INFO  (non-blocking-thread--p2-t22) [org.infinispan.CLUSTER] [Context=caching-listen]ISPN100002: Starting rebalance with members [jbolina-43637, jbolina-61891], phase READ_OLD_WRITE_ALL, topology id 2
13:57:03,994 INFO  (non-blocking-thread--p2-t22) [org.infinispan.LIFECYCLE] [Context=caching-listen][Scope=jbolina-61891]ISPN100002: Starting rebalance with members [jbolina-43637, jbolina-61891], phase READ_OLD_WRITE_ALL, topology id 2
```

jabolina commented:
Missed the update in the test, sorry.

ryanemerson commented:
Thanks @jabolina
